File: sportsbook-app/betting-service/src/main/java/com/sportsbook/bettingservice/messaging/BetStatusListener.java
package com.sportsbook.bettingservice.messaging;

import com.sportsbook.bettingservice.domain.BetStatus;
import com.sportsbook.bettingservice.repo.BetRepository;
import com.sportsbook.events.BetAccepted;
import com.sportsbook.events.BetRejected;
import com.sportsbook.events.BetSettled;
import lombok.RequiredArgsConstructor;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@RequiredArgsConstructor
public class BetStatusListener {

  private final BetRepository betRepository;

  @KafkaListener(topics = "bet.accepted", groupId = "betting-service")
  @Transactional
  public void onAccepted(BetAccepted evt) {
    betRepository.findById(evt.getBetId())
        .ifPresent(b -> {
          if (b.getStatus() == BetStatus.PENDING) {
            b.setStatus(BetStatus.ACCEPTED);
            betRepository.save(b);
          }
        });
  }

  @KafkaListener(topics = "bet.rejected", groupId = "betting-service")
  @Transactional
  public void onRejected(BetRejected evt) {
    betRepository.findById(evt.getBetId())
        .ifPresent(b -> {
          b.setStatus(BetStatus.REJECTED);
          betRepository.save(b);
        });
  }

  @KafkaListener(topics = "bet.settled", groupId = "betting-service")
  @Transactional
  public void onSettled(BetSettled evt) {
    betRepository.findById(evt.getBetId())
        .ifPresent(b -> {
          // If BetStatus has no WON/LOST, use SETTLED.
          b.setStatus(BetStatus.SETTLED);
          betRepository.save(b);
        });
  }
}

File: sportsbook-app/betting-service/src/main/kotlin/com/sportsbook/bettingservice/messaging/BetStatusListener.kt
package com.sportsbook.bettingservice.messaging

import com.sportsbook.bettingservice.domain.BetStatus
import com.sportsbook.bettingservice.repo.BetRepository
import com.sportsbook.events.BetAccepted
import com.sportsbook.events.BetRejected
import com.sportsbook.events.BetSettled
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class BetStatusListener(
    private val betRepository: BetRepository,
) {

    @KafkaListener(topics = ["bet.accepted"], groupId = "betting-service")
    @Transactional
    fun onAccepted(evt: BetAccepted) {
        betRepository.findById(evt.betId)
            .ifPresent { bet ->
                if (bet.status == BetStatus.PENDING) {
                    bet.status = BetStatus.ACCEPTED
                    betRepository.save(bet)
                }
            }
    }

    @KafkaListener(topics = ["bet.rejected"], groupId = "betting-service")
    @Transactional
    fun onRejected(evt: BetRejected) {
        betRepository.findById(evt.betId)
            .ifPresent { bet ->
                bet.status = BetStatus.REJECTED
                betRepository.save(bet)
            }
    }

    @KafkaListener(topics = ["bet.settled"], groupId = "betting-service")
    @Transactional
    fun onSettled(evt: BetSettled) {
        betRepository.findById(evt.betId)
            .ifPresent { bet ->
                bet.status = BetStatus.SETTLED
                betRepository.save(bet)
            }
    }
}
